src/signature.c

```diff
@@ -24,198 +24,379 @@
+	munmap(data, data_len);
+	close(data_fd);
+	BIO_free(sig_BIO);
+	BIO_free(data_BIO);
+	BIO_free(verify_BIO);
+	PKCS7_free(p7);
+
+	terminate_signature();
+
+	if (ret == 1) {
+		printf("Signature check succeeded.\n");
+		return true;
+	}
+
+	ERR_print_errors_fp(stderr);
+	fprintf(stderr, "Signature check failed!\n");
+	return false;
 }
 
-bool signature_verify(const char *data_filename, const char *sig_filename)
+/* Make sure the certificate exists and extract the public key from it.
+ *
+ * returns: true if it can get the public key, false otherwise */
+static bool get_pubkey(void)
 {
-	BIO *bio_data = NULL;
-	BIO *bio_sig = NULL;
-	PKCS7 *pkcs7 = NULL;
-	int ret;
-	bool result = false;
-
-	if (!initialized) {
-		return false || (IMPL == IMPL_FORGIVE);
-	}
-	bio_data = BIO_new_file(data_filename, "r"); // i.e. fopen
-	if (bio_data == NULL) {
-		goto exit;
-	}
-	bio_sig = BIO_new_file(sig_filename, "r"); // i.e. fopen
-	if (bio_sig == NULL) {
-		goto exit;
-	}
-	pkcs7 = PEM_read_bio_PKCS7(bio_sig, NULL, NULL, NULL);
-	if (pkcs7 == NULL) {
-		goto exit;
-	}
-	ret = PKCS7_verify(pkcs7, NULL, x509_store, bio_data, NULL, 0);
-	if (ret != 1) {
-		goto exit;
-	}
-	result = true;
-exit:
-	/*
-	 * The free functions below tolerate NULL arguments.
-	 * The documentation doesn't really say so, but both testing and
-	 * examination of openssl source code confirm that such is the case.
-         */
-	PKCS7_free(pkcs7);  // undocumented...
-	BIO_free(bio_sig);  // i.e. fclose
-	BIO_free(bio_data); // i.e. fclose
-	return result || (IMPL == IMPL_FORGIVE);
+	fp_pubkey = fopen(CERTNAME, "r");
+	if (!fp_pubkey) {
+		fprintf(stderr, "Failed fopen %s\n", CERTNAME);
+		goto error;
+	}
+
+	cert = PEM_read_X509(fp_pubkey, NULL, NULL, NULL);
+	if (!cert) {
+		fclose(fp_pubkey);
+		goto error;
+	}
+
+	pkey = X509_get_pubkey(cert);
+	if (!pkey) {
+		fclose(fp_pubkey);
+		X509_free(cert);
+		goto error;
+	}
+	return true;
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
 }
 
-static X509_STORE *create_store(const char *ca_filename, const char *ca_dirname,
-				const char *crl_filename)
+/* This function makes sure the certificate is still valid by not having any
+ * compromised certificates in the chain.
+ * If there is no Certificate Revocation List (CRL) it may be that the private
+ * keys have not been compromised or the CRL has not been generated by the
+ * Certificate Authority (CA)
+ *
+ * returns: true if certificate is valid, false otherwise */
+static bool validate_certificate(void)
 {
-	X509_STORE *store = X509_STORE_new();
+	X509_LOOKUP *lookup = NULL;
+	X509_STORE_CTX *verify_ctx = NULL;
 
-	if (!store) {
-		return NULL;
+	/* TODO: CRL and Chains are not required for the current setup, but we may
+	 * implement them in the future 
+	if (!crl) {
+		printf("No certificate revocation list provided\n");
 	}
-	if (X509_STORE_load_locations(store, ca_filename, ca_dirname) != 1) {
-		goto err;
+	if (!chain) {
+		printf("No certificate chain provided\n");
```

Comment:
Could reword this to: "Proceeding with signature validate using self-signed certificate"